Fix paligemma missing test_liger_kernel_to_instance_for_paligemma_instance test in test_monkey_patch.py

vvvdwbvvv opened the pull request:
## Summary
<!--- This is a required section; please describe the main purpose of this proposed code change. --->
WHAT THIS PR ADDS
- A new unit test file containing `test_apply_liger_kernel_to_instance_for_paligemma()`. Fix #776 
- The test instantiates a dummy PaliGemmaForConditionalGeneration model, confirms it is un-patched, runs `_apply_liger_kernel_to_instance()`, then verifies that:
  - model.forward is replaced by `paligemma_lce_forward()`.
  -  `vision_tower.vision_model.post_layernorm.forward` is replaced by `LigerLayerNorm.forward`.
  - Every encoder layer’s `layer_norm1.forward` and `layer_norm2.forward` are also replaced.
  - Source equality is checked with inspect.getsource before and after patching.

<!---
## Details
This is an optional section; is there anything specific that reviewers should be aware of?
--->

## Testing Done
<!--- This is a required section; please describe how this change was tested. --->

<!-- 
Replace BLANK with your device type. For example, A100-80G-PCIe

Complete the following tasks before sending your PR, and replace `[ ]` with
`[x]` to indicate you have done them. 
-->
`transformers==4.49.0`
<details>
  <summary>Test result</summary>

    
    ❯ python3 -m pytest test/transformers/test_monkey_patch.py -k paligemma -v -rP
    ============================================== test session starts ==============================================
    platform linux -- Python 3.11.11, pytest-8.4.1, pluggy-1.6.0 -- /home/vvvdwbvvv/.local/bin/python3
    cachedir: .pytest_cache
    rootdir: /home/vvvdwbvvv/develop/Liger-Kernel
    configfile: pyproject.toml
    plugins: asyncio-1.0.0
    asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
    collected 33 items / 32 deselected / 1 selected                                                                 
    
    test/transformers/test_monkey_patch.py::test_apply_liger_kernel_to_instance_for_paligemma 
    ------------------------------------------------- live log call -------------------------------------------------
    INFO     liger_kernel.transformers.monkey_patch:monkey_patch.py:1864 Applying Liger kernels to model instance with model type: paligemma with kwargs: {}
    PASSED                                                                                                    [100%]
    
    ==================================================== PASSES =====================================================
    _______________________________ test_apply_liger_kernel_to_instance_for_paligemma _______________________________
    --------------------------------------------- Captured stdout call ----------------------------------------------
    PaliGemmaForConditionalGeneration(
      (vision_tower): SiglipVisionModel(
        (vision_model): SiglipVisionTransformer(
          (embeddings): SiglipVisionEmbeddings(
            (patch_embedding): Conv2d(3, 48, kernel_size=(16, 16), stride=(16, 16), padding=valid)
            (position_embedding): Embedding(196, 48)
          )
          (encoder): SiglipEncoder(
            (layers): ModuleList(
              (0-1): 2 x SiglipEncoderLayer(
                (self_attn): SiglipSdpaAttention(
                  (k_proj): Linear(in_features=48, out_features=48, bias=True)
                  (v_proj): Linear(in_features=48, out_features=48, bias=True)
                  (q_proj): Linear(in_features=48, out_features=48, bias=True)
                  (out_proj): Linear(in_features=48, out_features=48, bias=True)
                )
                (layer_norm1): LigerLayerNorm((48,), eps=1e-05)
                (mlp): SiglipMLP(
                  (activation_fn): PytorchGELUTanh()
                  (fc1): Linear(in_features=48, out_features=64, bias=True)
                  (fc2): Linear(in_features=64, out_features=48, bias=True)
                )
                (layer_norm2): LigerLayerNorm((48,), eps=1e-05)
              )
            )
          )
          (post_layernorm): LigerLayerNorm((48,), eps=1e-05)
          (head): SiglipMultiheadAttentionPoolingHead(
            (attention): MultiheadAttention(
              (out_proj): NonDynamicallyQuantizableLinear(in_features=48, out_features=48, bias=True)
            )
            (layernorm): LigerLayerNorm((48,), eps=1e-05, elementwise_affine=True)
            (mlp): SiglipMLP(
              (activation_fn): PytorchGELUTanh()
              (fc1): Linear(in_features=48, out_features=64, bias=True)
              (fc2): Linear(in_features=64, out_features=48, bias=True)
            )
          )
        )
      )
      (multi_modal_projector): PaliGemmaMultiModalProjector(
        (linear): Linear(in_features=48, out_features=2048, bias=True)
      )
      (language_model): GemmaForCausalLM(
        (model): GemmaModel(
          (embed_tokens): Embedding(256000, 32, padding_idx=0)
          (layers): ModuleList(
            (0-1): 2 x GemmaDecoderLayer(
              (self_attn): GemmaAttention(
                (q_proj): Linear(in_features=32, out_features=4096, bias=False)
                (k_proj): Linear(in_features=32, out_features=4096, bias=False)
                (v_proj): Linear(in_features=32, out_features=4096, bias=False)
                (o_proj): Linear(in_features=4096, out_features=32, bias=False)
              )
              (mlp): LigerGEGLUMLP(
                (gate_proj): Linear(in_features=32, out_features=64, bias=False)
                (up_proj): Linear(in_features=32, out_features=64, bias=False)
                (down_proj): Linear(in_features=64, out_features=32, bias=False)
                (act_fn): SiLU()
              )
              (input_layernorm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
              (post_attention_layernorm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
            )
          )
          (norm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
          (rotary_emb): GemmaRotaryEmbedding()
        )
        (lm_head): Linear(in_features=32, out_features=256000, bias=False)
      )
    )
    ----------------------------------------------- Captured log call -----------------------------------------------
    INFO     liger_kernel.transformers.monkey_patch:monkey_patch.py:1864 Applying Liger kernels to model instance with model type: paligemma with kwargs: {}
    ======================================= 1 passed, 32 deselected in 1.78s ========================================
    
</details>

`transformers==4.53.1`

<details>
  <summary>Test result</summary>

    ❯ python3 -m pytest test/transformers/test_monkey_patch.py -k paligemma -v -rP
    ============================================== test session starts ==============================================
    platform linux -- Python 3.11.11, pytest-8.4.1, pluggy-1.6.0 -- /home/vvvdwbvvv/.local/bin/python3
    cachedir: .pytest_cache
    rootdir: /home/vvvdwbvvv/develop/Liger-Kernel
    configfile: pyproject.toml
    plugins: asyncio-1.0.0
    asyncio: mode=Mode.AUTO, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
    collected 33 items / 32 deselected / 1 selected                                                                 
    
    test/transformers/test_monkey_patch.py::test_apply_liger_kernel_to_instance_for_paligemma 
    ------------------------------------------------- live log call -------------------------------------------------
    INFO     liger_kernel.transformers.monkey_patch:monkey_patch.py:1864 Applying Liger kernels to model instance with model type: paligemma with kwargs: {}
    PASSED                                                                                                    [100%]
    
    ==================================================== PASSES =====================================================
    _______________________________ test_apply_liger_kernel_to_instance_for_paligemma _______________________________
    --------------------------------------------- Captured stdout call ----------------------------------------------
    PaliGemmaForConditionalGeneration(
      (model): PaliGemmaModel(
        (vision_tower): SiglipVisionModel(
          (vision_model): SiglipVisionTransformer(
            (embeddings): SiglipVisionEmbeddings(
              (patch_embedding): Conv2d(3, 48, kernel_size=(16, 16), stride=(16, 16), padding=valid)
              (position_embedding): Embedding(196, 48)
            )
            (encoder): SiglipEncoder(
              (layers): ModuleList(
                (0-1): 2 x SiglipEncoderLayer(
                  (layer_norm1): LigerLayerNorm((48,), eps=1e-05)
                  (self_attn): SiglipAttention(
                    (k_proj): Linear(in_features=48, out_features=48, bias=True)
                    (v_proj): Linear(in_features=48, out_features=48, bias=True)
                    (q_proj): Linear(in_features=48, out_features=48, bias=True)
                    (out_proj): Linear(in_features=48, out_features=48, bias=True)
                  )
                  (layer_norm2): LigerLayerNorm((48,), eps=1e-05)
                  (mlp): SiglipMLP(
                    (activation_fn): PytorchGELUTanh()
                    (fc1): Linear(in_features=48, out_features=64, bias=True)
                    (fc2): Linear(in_features=64, out_features=48, bias=True)
                  )
                )
              )
            )
            (post_layernorm): LigerLayerNorm((48,), eps=1e-05)
            (head): SiglipMultiheadAttentionPoolingHead(
              (attention): MultiheadAttention(
                (out_proj): NonDynamicallyQuantizableLinear(in_features=48, out_features=48, bias=True)
              )
              (layernorm): LigerLayerNorm((48,), eps=1e-05, elementwise_affine=True)
              (mlp): SiglipMLP(
                (activation_fn): PytorchGELUTanh()
                (fc1): Linear(in_features=48, out_features=64, bias=True)
                (fc2): Linear(in_features=64, out_features=48, bias=True)
              )
            )
          )
        )
        (multi_modal_projector): PaliGemmaMultiModalProjector(
          (linear): Linear(in_features=48, out_features=2048, bias=True)
        )
        (language_model): GemmaModel(
          (embed_tokens): Embedding(256000, 32, padding_idx=0)
          (layers): ModuleList(
            (0-1): 2 x GemmaDecoderLayer(
              (self_attn): GemmaAttention(
                (q_proj): Linear(in_features=32, out_features=4096, bias=False)
                (k_proj): Linear(in_features=32, out_features=4096, bias=False)
                (v_proj): Linear(in_features=32, out_features=4096, bias=False)
                (o_proj): Linear(in_features=4096, out_features=32, bias=False)
              )
              (mlp): LigerGEGLUMLP(
                (gate_proj): Linear(in_features=32, out_features=64, bias=False)
                (up_proj): Linear(in_features=32, out_features=64, bias=False)
                (down_proj): Linear(in_features=64, out_features=32, bias=False)
                (act_fn): SiLU()
              )
              (input_layernorm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
              (post_attention_layernorm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
            )
          )
          (norm): LigerRMSNorm((32,), eps=1e-05, offset=1.0, in_place=True, row_mode=None)
          (rotary_emb): GemmaRotaryEmbedding()
        )
      )
      (lm_head): Linear(in_features=32, out_features=256000, bias=False)
    )
    ----------------------------------------------- Captured log call -----------------------------------------------
    INFO     liger_kernel.transformers.monkey_patch:monkey_patch.py:1864 Applying Liger kernels to model instance with model type: paligemma with kwargs: {}
    ======================================= 1 passed, 32 deselected in 2.42s ========================================
</details>


- Hardware Type: <BLANK>
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [ ] run `make test-convergence` to ensure convergence

Tcc0403 commented:
Thanks for the contribution!

However, the test shows our current monkey patch for paligemma instance is not compatible with the latest transformers version.
In `transformers==4.52.4`, the test would fail with the following error message:

```bash
E               TypeError: The language_model of a PaliGemma model must be either GemmaForCausalLM or Gemma2ForCausalLM.
src/liger_kernel/transformers/monkey_patch.py:1171: TypeError
```

I think it is related to transformers VLM refactoring. See also: #723 


The error is raised due to the model class change in recent transformers versions (>=4.52.0)
https://github.com/linkedin/Liger-Kernel/blob/08906d3d9cad77fcced2b9f6db0a385eff154f03/src/liger_kernel/transformers/monkey_patch.py#L1170-L1173

Would you like to fix this issue as well?

vvvdwbvvv commented:
Yes, I would like to fix this issue as well. Does the fix be done in this PR as well?

Tcc0403 commented:
Yes, let's do it in this PR so we can check the fix with the test as well without breaking main branch ci.

vvvdwbvvv commented:
Fixed in 5b7f103 with `or isinstance(language_model, Gemma2Model)` for `transformer>=4.52.0` changes model class to from ```(language_model): GemmaForCausalLM( (model): GemmaModel())``` to `(language_model): GemmaModel()`

Tcc0403 commented on src/liger_kernel/transformers/monkey_patch.py:
```diff
@@ -1155,7 +1157,7 @@ def apply_liger_kernel_to_paligemma(
         language_model = model.language_model
 
-        if isinstance(language_model, GemmaForCausalLM):
+        if isinstance(language_model, GemmaForCausalLM) or isinstance(language_model, GemmaModel):
```
nit: [`isinstace(object, classinfo)`](https://docs.python.org/3/library/functions.html#isinstance) accepts *classinfo* as a tuple
```suggestion
        if isinstance(language_model, (GemmaForCausalLM, GemmaModel)):
```

Tcc0403 commented on src/liger_kernel/transformers/monkey_patch.py:
```diff
@@ -1165,7 +1167,7 @@ def apply_liger_kernel_to_paligemma(
             )
 
-        elif isinstance(language_model, Gemma2ForCausalLM):
+        elif isinstance(language_model, Gemma2ForCausalLM) or isinstance(language_model, Gemma2Model):
```
ditto

Tcc0403 commented on test/transformers/test_monkey_patch.py:
```diff
@@ -793,6 +804,59 @@ def test_apply_liger_kernel_to_instance_for_gemma2():
+        from transformers.models.paligemma.modeling_paligemma import PaliGemmaForConditionalGeneration
+
+        # Instantiate a dummy model
+        config = transformers.models.paligemma.configuration_paligemma.PaliGemmaConfig(
+            torch_dtype=torch.bfloat16,
+            rms_norm_eps=1e-5,
+            hidden_size=32,
+            intermediate_size=64,
+            hidden_act="silu",
+            num_hidden_layers=2,
+            vision_config={
+                "layer_norm_eps": 1e-5,
+                "hidden_size": 48,
+                "intermediate_size": 64,
+            },
+        )
```
`num_hidden_layers` wasn't properly set. It has to be wrapped in a dict as text_config 

ref: [PlaiGemmaConfig](https://github.com/huggingface/transformers/blob/ca7e1a3756c022bf31429c452b2f313f043f32de/src/transformers/models/paligemma/configuration_paligemma.py#L24) 

<details>
<summary>model with current config</summary>

```
❯ python3 -m pytest test/transformers/test_monkey_patch.py -k paligemma -v -rP
================================================= test session starts ==================================================
platform linux -- Python 3.10.12, pytest-8.3.4, pluggy-1.5.0 -- /home/tcc/Liger-Kernel/.venv/bin/python3
cachedir: .pytest_cache
rootdir: /home/tcc/Liger-Kernel
configfile: pyproject.toml
plugins: rerunfailures-15.0, xdist-3.6.1
collected 33 items / 32 deselected / 1 selected

test/transformers/test_monkey_patch.py::test_apply_liger_kernel_to_instance_for_paligemma
---------------------------------------------------- live log call -----------------------------------------------------
INFO     liger_kernel.transformers.monkey_patch:monkey_patch.py:1864 Applying Liger kernels to model instance with model type: paligemma with kwargs: {}
PASSED                                                                                                           [100%]

=================================================== warnings summary ===================================================
.venv/lib/python3.10/site-packages/_pytest/config/__init__.py:1441
  /home/tcc/Liger-Kernel/.venv/lib/python3.10/site-packages/_pytest/config/__init__.py:1441: PytestConfigWarning: Unknown config option: asyncio_mode

    self._warn_or_fail_if_strict(f"Unknown config option: {key}\n")

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================== PASSES ========================================================
__________________________________ test_apply_liger_kernel_to_instance_for_paligemma ___________________________________
------------------------------------------------- Captured stdout call -------------------------------------------------
PaliGemmaForConditionalGeneration(
  (model): PaliGemmaModel(
    (vision_tower): SiglipVisionModel(
      (vision_model): SiglipVisionTransformer(
        (embeddings): SiglipVisionEmbeddings(
          (patch_embedding): Conv2d(3, 48, kernel_size=(16, 16), stride=(16, 16), padding=valid)
          (position_embedding): Embedding(196, 48)
        )
        (encoder): SiglipEncoder(
          (layers): ModuleList(
            (0-11): 12 x SiglipEncoderLayer(
              (layer_norm1): LigerLayerNorm((48,), eps=1e-05)
              (self_attn): SiglipAttention(
                (k_proj): Linear(in_features=48, out_features=48, bias=True)
                (v_proj): Linear(in_features=48, out_features=48, bias=True)
                (q_proj): Linear(in_features=48, out_features=48, bias=True)
                (out_proj): Linear(in_features=48, out_features=48, bias=True)
              )
              (layer_norm2): LigerLayerNorm((48,), eps=1e-05)
              (mlp): SiglipMLP(
                (activation_fn): PytorchGELUTanh()
                (fc1): Linear(in_features=48, out_features=64, bias=True)
                (fc2): Linear(in_features=64, out_features=48, bias=True)
              )
            )
          )
        )
        (post_layernorm): LigerLayerNorm((48,), eps=1e-05)
        (head): SiglipMultiheadAttentionPoolingHead(
          (attention): MultiheadAttention(
            (out_proj): NonDynamicallyQuantizableLinear(in_features=48, out_features=48, bias=True)
          )
          (layernorm): LigerLayerNorm((48,), eps=1e-05, elementwise_affine=True)
          (mlp): SiglipMLP(
            (activation_fn): PytorchGELUTanh()
            (fc1): Linear(in_features=48, out_features=64, bias=True)
            (fc2): Linear(in_features=64, out_features=48, bias=True)
          )
        )
      )
    )
    (multi_modal_projector): PaliGemmaMultiModalProjector(
      (linear): Linear(in_features=48, out_features=2048, bias=True)
    )
    (language_model): GemmaModel(
      (embed_tokens): Embedding(257152, 2048, padding_idx=0)
      (layers): ModuleList(
        (0-17): 18 x GemmaDecoderLayer(
          (self_attn): GemmaAttention(
            (q_proj): Linear(in_features=2048, out_features=2048, bias=False)
            (k_proj): Linear(in_features=2048, out_features=256, bias=False)
            (v_proj): Linear(in_features=2048, out_features=256, bias=False)
            (o_proj): Linear(in_features=2048, out_features=2048, bias=False)
          )
          (mlp): LigerGEGLUMLP(
            (gate_proj): Linear(in_features=2048, out_features=16384, bias=False)
            (up_proj): Linear(in_features=2048, out_features=16384, bias=False)
            (down_proj): Linear(in_features=16384, out_features=2048, bias=False)
            (act_fn): PytorchGELUTanh()
          )
          (input_layernorm): LigerRMSNorm((2048,), eps=1e-06, offset=1.0, in_place=True, row_mode=None)
          (post_attention_layernorm): LigerRMSNorm((2048,), eps=1e-06, offset=1.0, in_place=True, row_mode=None)
        )
      )
      (norm): LigerRMSNorm((2048,), eps=1e-06, offset=1.0, in_place=True, row_mode=None)
      (rotary_emb): GemmaRotaryEmbedding()
    )
  )
  (lm_head): Linear(in_features=2048, out_features=257152, bias=False)
)
```

</details>

Tcc0403 replied:
```suggestion
        config = transformers.models.paligemma.configuration_paligemma.PaliGemmaConfig(
            torch_dtype=torch.bfloat16,
            text_config={
            	"num_hidden_layers": 2,
            	"rms_norm_eps": 1e-5,
	    	"hidden_size": 32,
	        "intermediate_size": 64,
	        "hidden_act": "silu",
            }, 
            vision_config={
            	"num_hidden_layers": 2,
                "layer_norm_eps": 1e-5,
                "hidden_size": 48,
                "intermediate_size": 64,
            },
        )
```

vvvdwbvvv commented:
Fixed in a393a50,  simplify with `(Gemma2ForCausalLM, Gemma2Model)` and modified the PaliGemmaConfig.
Test result of `transformer==4.49.0` and `transformer==4.53.1` is enclosed

Tcc0403 reviewed (changes requested):
There's a coding style issue. Run `make checkstsyle`.

vvvdwbvvv commented:
Formatted in 538ee26

Tcc0403 reviewed (approved):
Thanks! Congrats on your first contribution!
